// webdriver-image-comparison/dist/clientSideScripts/getElementPositionTopWindow.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var vitest_1 = require("vitest");
var getElementPositionTopWindow_js_1 = __importDefault(require("./getElementPositionTopWindow.js"));
(0, vitest_1.describe)('getElementPositionTopWindow', function () {
    (0, vitest_1.it)('should the the element position to the top of the window', function () {
        Element.prototype.getBoundingClientRect = vitest_1.vi.fn(function () {
            return {
                width: 120,
                height: 120,
                top: 10,
                left: 100,
                bottom: 5,
                right: 12,
            };
        });
        document.body.innerHTML = '<div>' + '  <span id="username">Hello</span>' + '</div>';
        (0, vitest_1.expect)((0, getElementPositionTopWindow_js_1.default)(document.querySelector('#username'))).toMatchSnapshot();
    });
});
