// webdriver-image-comparison/dist/helpers/options.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var vitest_1 = require("vitest");
var options_js_1 = require("./options.js");
(0, vitest_1.describe)('options', function () {
    (0, vitest_1.describe)('defaultOptions', function () {
        (0, vitest_1.it)('should return the default options when no options are provided', function () {
            (0, vitest_1.expect)((0, options_js_1.defaultOptions)({})).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return the provided options when options are provided', function () {
            var options = {
                addressBarShadowPadding: 1,
                autoSaveBaseline: true,
                formatImageName: '{foo}-{bar}',
                savePerInstance: true,
                toolBarShadowPadding: 1,
                disableCSSAnimation: true,
                fullPageScrollTimeout: 12345,
                hideScrollBars: true,
                blockOutSideBar: true,
                blockOutStatusBar: true,
                blockOutToolBar: true,
                ignoreAlpha: true,
                ignoreAntialiasing: true,
                ignoreColors: true,
                ignoreLess: true,
                ignoreNothing: true,
                rawMisMatchPercentage: true,
                returnAllCompareData: true,
                saveAboveTolerance: 12,
                scaleImagesToSameSize: true,
                tabbableOptions: {
                    circle: {
                        backgroundColor: 'backgroundColor',
                        borderColor: 'borderColor',
                        borderWidth: 123,
                        fontColor: 'fontColor',
                        fontFamily: 'fontFamily',
                        fontSize: 321,
                        size: 567,
                        showNumber: false,
                    },
                    line: {
                        color: 'color',
                        width: 987,
                    },
                },
            };
            (0, vitest_1.expect)((0, options_js_1.defaultOptions)(options)).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('methodCompareOptions', function () {
        (0, vitest_1.it)('should not return the method options when no options are provided', function () {
            (0, vitest_1.expect)((0, options_js_1.methodCompareOptions)({})).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return the provided options when options are provided', function () {
            var options = {
                blockOut: [{ height: 1, width: 2, x: 3, y: 4 }],
                ignoreAlpha: true,
                ignoreAntialiasing: true,
                ignoreColors: true,
                ignoreLess: true,
                ignoreNothing: true,
                rawMisMatchPercentage: true,
                returnAllCompareData: true,
                saveAboveTolerance: 12,
                scaleImagesToSameSize: true,
            };
            (0, vitest_1.expect)((0, options_js_1.methodCompareOptions)(options)).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('screenMethodCompareOptions', function () {
        (0, vitest_1.it)('should not return the screen method options when no options are provided', function () {
            (0, vitest_1.expect)((0, options_js_1.screenMethodCompareOptions)({})).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return the provided options when options are provided', function () {
            var options = {
                blockOutSideBar: false,
                blockOutStatusBar: false,
                blockOutToolBar: false,
                blockOut: [{ height: 1, width: 2, x: 3, y: 4 }],
                ignoreAlpha: true,
                ignoreAntialiasing: true,
                ignoreColors: true,
                ignoreLess: true,
                ignoreNothing: true,
                rawMisMatchPercentage: true,
                returnAllCompareData: true,
                saveAboveTolerance: 12,
                scaleImagesToSameSize: true,
            };
            (0, vitest_1.expect)((0, options_js_1.screenMethodCompareOptions)(options)).toMatchSnapshot();
        });
    });
});
