// webdriver-image-comparison/dist/helpers/options.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.methodCompareOptions = exports.screenMethodCompareOptions = exports.defaultOptions = void 0;
var constants_js_1 = require("./constants.js");
var options_interfaces_1 = require("./options.interfaces");
function defaultOptions(options) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0;
    return {
        addressBarShadowPadding: (_a = options.addressBarShadowPadding) !== null && _a !== void 0 ? _a : constants_js_1.DEFAULT_SHADOW.ADDRESS_BAR,
        autoElementScroll: Object.prototype.hasOwnProperty.call(options, 'autoElementScroll')
            ? Boolean(options.autoElementScroll)
            : true,
        addIOSBezelCorners: (_b = options.addIOSBezelCorners) !== null && _b !== void 0 ? _b : false,
        autoSaveBaseline: (_c = options.autoSaveBaseline) !== null && _c !== void 0 ? _c : true,
        clearFolder: (_d = options.clearRuntimeFolder) !== null && _d !== void 0 ? _d : false,
        formatImageName: (_e = options.formatImageName) !== null && _e !== void 0 ? _e : constants_js_1.DEFAULT_FORMAT_STRING,
        isHybridApp: (_f = options.isHybridApp) !== null && _f !== void 0 ? _f : false,
        logLevel: (_g = options.logLevel) !== null && _g !== void 0 ? _g : options_interfaces_1.LogLevel.info,
        savePerInstance: (_h = options.savePerInstance) !== null && _h !== void 0 ? _h : false,
        toolBarShadowPadding: (_j = options.toolBarShadowPadding) !== null && _j !== void 0 ? _j : constants_js_1.DEFAULT_SHADOW.TOOL_BAR,
        disableCSSAnimation: (_k = options.disableCSSAnimation) !== null && _k !== void 0 ? _k : false,
        enableLayoutTesting: (_l = options.enableLayoutTesting) !== null && _l !== void 0 ? _l : false,
        fullPageScrollTimeout: (_m = options.fullPageScrollTimeout) !== null && _m !== void 0 ? _m : constants_js_1.FULL_PAGE_SCROLL_TIMEOUT,
        hideScrollBars: Object.prototype.hasOwnProperty.call(options, 'hideScrollBars')
            ? Boolean(options.hideScrollBars)
            : true,
        waitForFontsLoaded: (_o = options.waitForFontsLoaded) !== null && _o !== void 0 ? _o : true,
        compareOptions: {
            blockOutSideBar: (_p = options.blockOutSideBar) !== null && _p !== void 0 ? _p : true,
            blockOutStatusBar: (_q = options.blockOutStatusBar) !== null && _q !== void 0 ? _q : true,
            blockOutToolBar: (_r = options.blockOutToolBar) !== null && _r !== void 0 ? _r : true,
            ignoreAlpha: (_s = options.ignoreAlpha) !== null && _s !== void 0 ? _s : false,
            ignoreAntialiasing: (_t = options.ignoreAntialiasing) !== null && _t !== void 0 ? _t : false,
            ignoreColors: (_u = options.ignoreColors) !== null && _u !== void 0 ? _u : false,
            ignoreLess: (_v = options.ignoreLess) !== null && _v !== void 0 ? _v : false,
            ignoreNothing: (_w = options.ignoreNothing) !== null && _w !== void 0 ? _w : false,
            rawMisMatchPercentage: (_x = options.rawMisMatchPercentage) !== null && _x !== void 0 ? _x : false,
            returnAllCompareData: (_y = options.returnAllCompareData) !== null && _y !== void 0 ? _y : false,
            saveAboveTolerance: (_z = options.saveAboveTolerance) !== null && _z !== void 0 ? _z : 0,
            scaleImagesToSameSize: (_0 = options.scaleImagesToSameSize) !== null && _0 !== void 0 ? _0 : false,
        },
        tabbableOptions: {
            circle: __assign(__assign({}, constants_js_1.DEFAULT_TABBABLE_OPTIONS.circle), (options.tabbableOptions && options.tabbableOptions.circle ? options.tabbableOptions.circle : {})),
            line: __assign(__assign({}, constants_js_1.DEFAULT_TABBABLE_OPTIONS.line), (options.tabbableOptions && options.tabbableOptions.line ? options.tabbableOptions.line : {})),
        },
    };
}
exports.defaultOptions = defaultOptions;
function screenMethodCompareOptions(options) {
    return __assign(__assign(__assign(__assign({}, ('blockOutSideBar' in options ? { blockOutSideBar: options.blockOutSideBar } : {})), ('blockOutStatusBar' in options ? { blockOutStatusBar: options.blockOutStatusBar } : {})), ('blockOutToolBar' in options ? { blockOutToolBar: options.blockOutToolBar } : {})), methodCompareOptions(options));
}
exports.screenMethodCompareOptions = screenMethodCompareOptions;
function methodCompareOptions(options) {
    return __assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign({}, ('blockOut' in options ? { blockOut: options.blockOut } : {})), ('ignoreAlpha' in options ? { ignoreAlpha: options.ignoreAlpha } : {})), ('ignoreAntialiasing' in options ? { ignoreAntialiasing: options.ignoreAntialiasing } : {})), ('ignoreColors' in options ? { ignoreColors: options.ignoreColors } : {})), ('ignoreLess' in options ? { ignoreLess: options.ignoreLess } : {})), ('ignoreNothing' in options ? { ignoreNothing: options.ignoreNothing } : {})), ('rawMisMatchPercentage' in options ? { rawMisMatchPercentage: options.rawMisMatchPercentage } : {})), ('returnAllCompareData' in options ? { returnAllCompareData: options.returnAllCompareData } : {})), ('saveAboveTolerance' in options ? { saveAboveTolerance: options.saveAboveTolerance } : {})), ('scaleImagesToSameSize' in options ? { scaleImagesToSameSize: options.scaleImagesToSameSize } : {}));
}
exports.methodCompareOptions = methodCompareOptions;
