// webdriver-image-comparison/dist/helpers/utils.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var vitest_1 = require("vitest");
var fs_extra_1 = require("fs-extra");
var node_path_1 = require("node:path");
var utils_js_1 = require("./utils.js");
var mocks_js_1 = require("../mocks/mocks.js");
(0, vitest_1.describe)('utils', function () {
    (0, vitest_1.describe)('getAndCreatePath', function () {
        var folder = (0, node_path_1.join)(process.cwd(), '/.tmp/utils');
        (0, vitest_1.afterEach)(function () { return (0, fs_extra_1.removeSync)(folder); });
        (0, vitest_1.it)('should create the folder and return the folder name for a device that needs to have its own folder', function () {
            var options = {
                browserName: '',
                deviceName: 'deviceName',
                isMobile: true,
                savePerInstance: true,
            };
            var expectedFolderName = (0, node_path_1.join)(folder, options.deviceName);
            (0, vitest_1.expect)((0, fs_extra_1.pathExistsSync)(expectedFolderName)).toMatchSnapshot();
            (0, vitest_1.expect)((0, utils_js_1.getAndCreatePath)(folder, options)).toEqual(expectedFolderName);
            (0, vitest_1.expect)((0, fs_extra_1.pathExistsSync)(expectedFolderName)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should create the folder and return the folder name for a browser that needs to have its own folder', function () {
            var options = {
                browserName: 'browser',
                deviceName: '',
                isMobile: false,
                savePerInstance: true,
            };
            var expectedFolderName = (0, node_path_1.join)(folder, "desktop_".concat(options.browserName));
            (0, vitest_1.expect)((0, fs_extra_1.pathExistsSync)(expectedFolderName)).toMatchSnapshot();
            (0, vitest_1.expect)((0, utils_js_1.getAndCreatePath)(folder, options)).toEqual(expectedFolderName);
            (0, vitest_1.expect)((0, fs_extra_1.pathExistsSync)(expectedFolderName)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should create the folder and return the folder name for a browser', function () {
            var options = {
                browserName: 'browser',
                deviceName: '',
                isMobile: false,
                savePerInstance: false,
            };
            (0, vitest_1.expect)((0, fs_extra_1.pathExistsSync)(folder)).toMatchSnapshot();
            (0, vitest_1.expect)((0, utils_js_1.getAndCreatePath)(folder, options)).toEqual(folder);
            (0, vitest_1.expect)((0, fs_extra_1.pathExistsSync)(folder)).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('formatFileName', function () {
        var formatFileOptions = {
            browserName: '',
            browserVersion: '',
            deviceName: '',
            devicePixelRatio: 2,
            formatImageName: '',
            isMobile: false,
            isTestInBrowser: true,
            logName: '',
            name: '',
            outerHeight: 768,
            outerWidth: 1366,
            platformName: '',
            platformVersion: '',
            screenHeight: 900,
            screenWidth: 1400,
            tag: 'theTag',
        };
        (0, vitest_1.it)('should format a string with all options provided', function () {
            formatFileOptions.formatImageName =
                'browser.{browserName}-{browserVersion}-platform.{platformName}-{platformVersion}-dpr.{dpr}-{height}-{logName}-{name}-{tag}-{width}';
            formatFileOptions.browserName = 'chrome';
            formatFileOptions.browserVersion = '74';
            formatFileOptions.logName = 'chrome-latest';
            formatFileOptions.name = 'chrome-name';
            formatFileOptions.platformName = 'osx';
            formatFileOptions.platformVersion = '12';
            (0, vitest_1.expect)((0, utils_js_1.formatFileName)(formatFileOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should format a string for mobile app', function () {
            formatFileOptions.formatImageName = '{tag}-{mobile}-{dpr}-{width}x{height}';
            formatFileOptions.deviceName = 'iPhoneX';
            formatFileOptions.isMobile = true;
            formatFileOptions.isTestInBrowser = false;
            (0, vitest_1.expect)((0, utils_js_1.formatFileName)(formatFileOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should format a string for mobile browser', function () {
            formatFileOptions.formatImageName = '{tag}-{mobile}-{dpr}-{width}x{height}';
            formatFileOptions.browserName = 'chrome';
            formatFileOptions.deviceName = 'iPhoneX';
            formatFileOptions.isMobile = true;
            formatFileOptions.isTestInBrowser = true;
            (0, vitest_1.expect)((0, utils_js_1.formatFileName)(formatFileOptions)).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('checkIsMobile', function () {
        (0, vitest_1.it)('should return false when no platform name is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkIsMobile)('')).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return true when a platform name is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkIsMobile)('ios')).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('checkIsAndroid', function () {
        (0, vitest_1.it)('should return false when no platform name is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkIsAndroid)('')).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return false when a platform name is provided that is not accepted', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkIsAndroid)('chrome')).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return true when a valid platform name is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkIsAndroid)('androId')).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('checkIsIos', function () {
        (0, vitest_1.it)('should return false when no platform name is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkIsIos)('')).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return false when a platform name is provided that is not accepted', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkIsIos)('chrome')).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return true when a valid platform name is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkIsIos)('IoS')).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('checkTestInBrowser', function () {
        (0, vitest_1.it)('should return false when no browser name is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkTestInBrowser)('')).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return true when a browser name is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkTestInBrowser)('chrome')).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('checkTestInMobileBrowser', function () {
        (0, vitest_1.it)('should return false when no platform name is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkTestInMobileBrowser)('', 'chrome')).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return false when a plaform but no browser name is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkTestInMobileBrowser)('ios', '')).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return true when a plaform and a browser name is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkTestInMobileBrowser)('ios', 'chrome')).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('checkAndroidNativeWebScreenshot', function () {
        (0, vitest_1.it)('should return false when no platform name is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkAndroidNativeWebScreenshot)('', false)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return false when iOS and nativeWebscreenshot true is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkAndroidNativeWebScreenshot)('ios', true)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return false when iOS and nativeWebscreenshot false is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkAndroidNativeWebScreenshot)('ios', false)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return false when Android and nativeWebscreenshot false is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkAndroidNativeWebScreenshot)('Android', false)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return true when Android and nativeWebscreenshot true is provided ', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkAndroidNativeWebScreenshot)('Android', true)).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('checkAndroidChromeDriverScreenshot', function () {
        (0, vitest_1.it)('should return false when no platform name is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkAndroidChromeDriverScreenshot)('', false)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return false when iOS and nativeWebscreenshot true is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkAndroidChromeDriverScreenshot)('ios', true)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return false when iOS and nativeWebscreenshot false is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkAndroidChromeDriverScreenshot)('ios', false)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return false when Android and nativeWebscreenshot true is provided', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkAndroidChromeDriverScreenshot)('Android', true)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return true when Android and nativeWebscreenshot false is provided ', function () {
            (0, vitest_1.expect)((0, utils_js_1.checkAndroidChromeDriverScreenshot)('Android', false)).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('getAddressBarShadowPadding', function () {
        var getAddressBarShadowPaddingOptions = {
            platformName: '',
            browserName: '',
            nativeWebScreenshot: false,
            addressBarShadowPadding: 6,
            addShadowPadding: false,
        };
        (0, vitest_1.it)('should return 0 when this is a check for a desktop browser', function () {
            getAddressBarShadowPaddingOptions.browserName = 'chrome';
            (0, vitest_1.expect)((0, utils_js_1.getAddressBarShadowPadding)(getAddressBarShadowPaddingOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return 0 when this is a check for an Android app', function () {
            getAddressBarShadowPaddingOptions.platformName = 'android';
            (0, vitest_1.expect)((0, utils_js_1.getAddressBarShadowPadding)(getAddressBarShadowPaddingOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return 0 when this is a check for an iOS app', function () {
            getAddressBarShadowPaddingOptions.platformName = 'ios';
            (0, vitest_1.expect)((0, utils_js_1.getAddressBarShadowPadding)(getAddressBarShadowPaddingOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return 0 when this is a check for Android with a native screenshot but without adding a shadow padding', function () {
            getAddressBarShadowPaddingOptions.platformName = 'android';
            getAddressBarShadowPaddingOptions.nativeWebScreenshot = true;
            (0, vitest_1.expect)((0, utils_js_1.getAddressBarShadowPadding)(getAddressBarShadowPaddingOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return 0 when this is a check for iOS but without adding a shadow padding', function () {
            getAddressBarShadowPaddingOptions.platformName = 'iOS';
            getAddressBarShadowPaddingOptions.nativeWebScreenshot = true;
            (0, vitest_1.expect)((0, utils_js_1.getAddressBarShadowPadding)(getAddressBarShadowPaddingOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return 6 when this is a check for Android with a native screenshot and adding a shadow padding', function () {
            getAddressBarShadowPaddingOptions.platformName = 'android';
            getAddressBarShadowPaddingOptions.nativeWebScreenshot = true;
            getAddressBarShadowPaddingOptions.addShadowPadding = true;
            (0, vitest_1.expect)((0, utils_js_1.getAddressBarShadowPadding)(getAddressBarShadowPaddingOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return 6 when this is a check for iOS and adding a shadow padding', function () {
            getAddressBarShadowPaddingOptions.platformName = 'iOS';
            getAddressBarShadowPaddingOptions.addShadowPadding = true;
            (0, vitest_1.expect)((0, utils_js_1.getAddressBarShadowPadding)(getAddressBarShadowPaddingOptions)).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('getToolBarShadowPadding', function () {
        (0, vitest_1.it)('should return 0 when this is a check for a desktop browser', function () {
            var getToolBarShadowPaddingOptions = {
                platformName: '',
                browserName: 'chrome',
                toolBarShadowPadding: 6,
                addShadowPadding: false,
            };
            (0, vitest_1.expect)((0, utils_js_1.getToolBarShadowPadding)(getToolBarShadowPaddingOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return 0 when this is a check for an Android app', function () {
            var getToolBarShadowPaddingOptions = {
                platformName: 'Android',
                browserName: '',
                toolBarShadowPadding: 6,
                addShadowPadding: false,
            };
            (0, vitest_1.expect)((0, utils_js_1.getToolBarShadowPadding)(getToolBarShadowPaddingOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return 0 when this is a check for an iOS app', function () {
            var getToolBarShadowPaddingOptions = {
                platformName: 'iOS',
                browserName: '',
                toolBarShadowPadding: 6,
                addShadowPadding: false,
            };
            (0, vitest_1.expect)((0, utils_js_1.getToolBarShadowPadding)(getToolBarShadowPaddingOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return 0 when this is a check for an Android app with adding a shadow padding', function () {
            var getToolBarShadowPaddingOptions = {
                platformName: 'android',
                browserName: '',
                toolBarShadowPadding: 6,
                addShadowPadding: true,
            };
            (0, vitest_1.expect)((0, utils_js_1.getToolBarShadowPadding)(getToolBarShadowPaddingOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return 0 when this is a check for an iOS app with adding a shadow padding', function () {
            var getToolBarShadowPaddingOptions = {
                platformName: 'iOS',
                browserName: '',
                toolBarShadowPadding: 6,
                addShadowPadding: true,
            };
            (0, vitest_1.expect)((0, utils_js_1.getToolBarShadowPadding)(getToolBarShadowPaddingOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return 0 when this is a check for Android browser and adding a shadow padding', function () {
            var getToolBarShadowPaddingOptions = {
                platformName: 'android',
                browserName: 'chrome',
                toolBarShadowPadding: 6,
                addShadowPadding: true,
            };
            (0, vitest_1.expect)((0, utils_js_1.getToolBarShadowPadding)(getToolBarShadowPaddingOptions)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should return 15 when this is a check for iOS browser and adding a shadow padding', function () {
            var getToolBarShadowPaddingOptions = {
                platformName: 'ios',
                browserName: 'safari',
                toolBarShadowPadding: 6,
                addShadowPadding: true,
            };
            (0, vitest_1.expect)((0, utils_js_1.getToolBarShadowPadding)(getToolBarShadowPaddingOptions)).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('calculateDprData', function () {
        (0, vitest_1.it)('should multiple all number values by the dpr value', function () {
            var data = {
                a: 1,
                b: 2,
                1: 3,
                a1: 9,
                bool: true,
                string: 'string',
            };
            (0, vitest_1.expect)((0, utils_js_1.calculateDprData)(data, 2)).toMatchSnapshot();
        });
    });
    (0, vitest_1.describe)('getScreenshotSize', function () {
        (0, vitest_1.it)('should get the screenshot size of a screenshot string with the default DPR', function () {
            (0, vitest_1.expect)((0, utils_js_1.getScreenshotSize)(mocks_js_1.IMAGE_STRING)).toMatchSnapshot();
        });
        (0, vitest_1.it)('should get the screenshot size of a screenshot string with DRP 2', function () {
            (0, vitest_1.expect)((0, utils_js_1.getScreenshotSize)(mocks_js_1.IMAGE_STRING, 2)).toMatchSnapshot();
        });
    });
});
