// webdriver-image-comparison/dist/commands/saveWebElement.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var screenshots_js_1 = require("../methods/screenshots.js");
var images_js_1 = require("../methods/images.js");
var beforeScreenshot_js_1 = __importDefault(require("../helpers/beforeScreenshot.js"));
var afterScreenshot_js_1 = __importDefault(require("../helpers/afterScreenshot.js"));
var rectangles_js_1 = require("../methods/rectangles.js");
var constants_js_1 = require("../helpers/constants.js");
var scrollElementIntoView_js_1 = __importDefault(require("../clientSideScripts/scrollElementIntoView.js"));
var utils_js_1 = require("../helpers/utils.js");
var scrollToPosition_js_1 = __importDefault(require("../clientSideScripts/scrollToPosition.js"));
function saveWebElement(methods, instanceData, folders, element, tag, saveElementOptions, isNativeContext) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, addressBarShadowPadding, autoElementScroll, formatImageName, logLevel, savePerInstance, toolBarShadowPadding, executor, disableCSSAnimation, enableLayoutTesting, hideScrollBars, resizeDimensions, hideElements, removeElements, waitForFontsLoaded, beforeOptions, enrichedInstanceData, browserName, browserVersion, deviceName, _b, devicePixelRatio, innerHeight, isLandscape, outerHeight, outerWidth, screenHeight, screenWidth, isAndroid, isAndroidNativeWebScreenshot, isIOS, isMobile, isTestInBrowser, logName, name, platformName, platformVersion, currentPosition, base64Image, elementRectangleOptions, rectangles, croppedBase64Image, afterOptions;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _a = saveElementOptions.wic, addressBarShadowPadding = _a.addressBarShadowPadding, autoElementScroll = _a.autoElementScroll, formatImageName = _a.formatImageName, logLevel = _a.logLevel, savePerInstance = _a.savePerInstance, toolBarShadowPadding = _a.toolBarShadowPadding;
                    executor = methods.executor;
                    disableCSSAnimation = saveElementOptions.method.disableCSSAnimation !== undefined
                        ? Boolean(saveElementOptions.method.disableCSSAnimation)
                        : saveElementOptions.wic.disableCSSAnimation;
                    enableLayoutTesting = saveElementOptions.method.enableLayoutTesting !== undefined
                        ? Boolean(saveElementOptions.method.enableLayoutTesting)
                        : saveElementOptions.wic.enableLayoutTesting;
                    hideScrollBars = saveElementOptions.method.hideScrollBars !== undefined
                        ? Boolean(saveElementOptions.method.hideScrollBars)
                        : saveElementOptions.wic.hideScrollBars;
                    resizeDimensions = saveElementOptions.method.resizeDimensions || constants_js_1.DEFAULT_RESIZE_DIMENSIONS;
                    hideElements = saveElementOptions.method.hideElements || [];
                    removeElements = saveElementOptions.method.removeElements || [];
                    waitForFontsLoaded = saveElementOptions.method.waitForFontsLoaded !== undefined
                        ? Boolean(saveElementOptions.method.waitForFontsLoaded)
                        : saveElementOptions.wic.waitForFontsLoaded;
                    beforeOptions = {
                        instanceData: instanceData,
                        addressBarShadowPadding: addressBarShadowPadding,
                        disableCSSAnimation: disableCSSAnimation,
                        enableLayoutTesting: enableLayoutTesting,
                        hideElements: hideElements,
                        logLevel: logLevel,
                        noScrollBars: hideScrollBars,
                        removeElements: removeElements,
                        toolBarShadowPadding: toolBarShadowPadding,
                        waitForFontsLoaded: waitForFontsLoaded,
                    };
                    return [4, (0, beforeScreenshot_js_1.default)(executor, beforeOptions, true)];
                case 1:
                    enrichedInstanceData = _c.sent();
                    browserName = enrichedInstanceData.browserName, browserVersion = enrichedInstanceData.browserVersion, deviceName = enrichedInstanceData.deviceName, _b = enrichedInstanceData.dimensions.window, devicePixelRatio = _b.devicePixelRatio, innerHeight = _b.innerHeight, isLandscape = _b.isLandscape, outerHeight = _b.outerHeight, outerWidth = _b.outerWidth, screenHeight = _b.screenHeight, screenWidth = _b.screenWidth, isAndroid = enrichedInstanceData.isAndroid, isAndroidNativeWebScreenshot = enrichedInstanceData.isAndroidNativeWebScreenshot, isIOS = enrichedInstanceData.isIOS, isMobile = enrichedInstanceData.isMobile, isTestInBrowser = enrichedInstanceData.isTestInBrowser, logName = enrichedInstanceData.logName, name = enrichedInstanceData.name, platformName = enrichedInstanceData.platformName, platformVersion = enrichedInstanceData.platformVersion;
                    if (!autoElementScroll) return [3, 4];
                    return [4, executor(scrollElementIntoView_js_1.default, element, addressBarShadowPadding)];
                case 2:
                    currentPosition = _c.sent();
                    return [4, (0, utils_js_1.waitFor)(500)];
                case 3:
                    _c.sent();
                    _c.label = 4;
                case 4: return [4, (0, screenshots_js_1.takeBase64Screenshot)(methods.screenShot)];
                case 5:
                    base64Image = _c.sent();
                    elementRectangleOptions = {
                        devicePixelRatio: devicePixelRatio || NaN,
                        innerHeight: innerHeight || NaN,
                        isAndroidNativeWebScreenshot: isAndroidNativeWebScreenshot,
                        isAndroid: isAndroid,
                        isIOS: isIOS,
                        isLandscape: isLandscape,
                    };
                    return [4, (0, rectangles_js_1.determineElementRectangles)({
                            executor: executor,
                            base64Image: base64Image,
                            options: elementRectangleOptions,
                            element: element,
                        })];
                case 6:
                    rectangles = _c.sent();
                    if (!(autoElementScroll && currentPosition)) return [3, 8];
                    return [4, executor(scrollToPosition_js_1.default, currentPosition)];
                case 7:
                    _c.sent();
                    _c.label = 8;
                case 8: return [4, (0, images_js_1.makeCroppedBase64Image)({
                        addIOSBezelCorners: false,
                        base64Image: base64Image,
                        deviceName: deviceName,
                        devicePixelRatio: devicePixelRatio || NaN,
                        isIOS: isIOS,
                        isLandscape: isLandscape,
                        logLevel: logLevel,
                        rectangles: rectangles,
                        resizeDimensions: resizeDimensions,
                    })];
                case 9:
                    croppedBase64Image = _c.sent();
                    afterOptions = {
                        actualFolder: folders.actualFolder,
                        base64Image: croppedBase64Image,
                        disableCSSAnimation: disableCSSAnimation,
                        enableLayoutTesting: enableLayoutTesting,
                        filePath: {
                            browserName: browserName,
                            deviceName: deviceName,
                            isMobile: isMobile,
                            savePerInstance: savePerInstance,
                        },
                        fileName: {
                            browserName: browserName,
                            browserVersion: browserVersion,
                            deviceName: deviceName,
                            devicePixelRatio: devicePixelRatio || NaN,
                            formatImageName: formatImageName,
                            isMobile: isMobile,
                            isTestInBrowser: isTestInBrowser,
                            logName: logName,
                            name: name,
                            outerHeight: outerHeight || NaN,
                            outerWidth: outerWidth || NaN,
                            platformName: platformName,
                            platformVersion: platformVersion,
                            screenHeight: screenHeight || NaN,
                            screenWidth: screenWidth || NaN,
                            tag: tag,
                        },
                        hideElements: hideElements,
                        hideScrollBars: hideScrollBars,
                        isLandscape: isLandscape,
                        isNativeContext: false,
                        logLevel: logLevel,
                        platformName: instanceData.platformName,
                        removeElements: removeElements,
                    };
                    return [2, (0, afterScreenshot_js_1.default)(executor, afterOptions)];
            }
        });
    });
}
exports.default = saveWebElement;
