// webdriver-image-comparison/dist/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var node_path_1 = require("node:path");
var fs_extra_1 = require("fs-extra");
var options_js_1 = require("./helpers/options.js");
var constants_js_1 = require("./helpers/constants.js");
var BaseClass = (function () {
    function BaseClass(options) {
        this.defaultOptions = (0, options_js_1.defaultOptions)(options);
        var baselineFolder = typeof options.baselineFolder === 'function'
            ? options.baselineFolder(options)
            : (0, node_path_1.normalize)(options.baselineFolder || constants_js_1.FOLDERS.DEFAULT.BASE);
        var baseFolder = typeof options.screenshotPath === 'function'
            ? options.screenshotPath(options)
            : (0, node_path_1.normalize)(options.screenshotPath || constants_js_1.FOLDERS.DEFAULT.SCREENSHOTS);
        this.folders = {
            actualFolder: (0, node_path_1.join)(baseFolder, constants_js_1.FOLDERS.ACTUAL),
            baselineFolder: baselineFolder,
            diffFolder: (0, node_path_1.join)(baseFolder, constants_js_1.FOLDERS.DIFF),
        };
        if (options.clearRuntimeFolder) {
            console.log('\n\n\n##############################');
            console.log('!!CLEARING!!');
            console.log('##############################\n\n\n');
            (0, fs_extra_1.removeSync)(this.folders.actualFolder);
            (0, fs_extra_1.removeSync)(this.folders.diffFolder);
        }
    }
    return BaseClass;
}());
exports.default = BaseClass;
