// webdriver-image-comparison/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkTabbablePage = exports.checkFullPageScreen = exports.checkElement = exports.checkScreen = exports.saveTabbablePage = exports.saveFullPageScreen = exports.saveElement = exports.saveScreen = exports.FOLDERS = exports.IOS_OFFSETS = exports.BaseClass = void 0;
var base_js_1 = __importDefault(require("./base.js"));
exports.BaseClass = base_js_1.default;
var saveScreen_js_1 = __importDefault(require("./commands/saveScreen.js"));
exports.saveScreen = saveScreen_js_1.default;
var saveElement_js_1 = __importDefault(require("./commands/saveElement.js"));
exports.saveElement = saveElement_js_1.default;
var saveFullPageScreen_js_1 = __importDefault(require("./commands/saveFullPageScreen.js"));
exports.saveFullPageScreen = saveFullPageScreen_js_1.default;
var saveTabbablePage_js_1 = __importDefault(require("./commands/saveTabbablePage.js"));
exports.saveTabbablePage = saveTabbablePage_js_1.default;
var checkScreen_js_1 = __importDefault(require("./commands/checkScreen.js"));
exports.checkScreen = checkScreen_js_1.default;
var checkElement_js_1 = __importDefault(require("./commands/checkElement.js"));
exports.checkElement = checkElement_js_1.default;
var checkFullPageScreen_js_1 = __importDefault(require("./commands/checkFullPageScreen.js"));
exports.checkFullPageScreen = checkFullPageScreen_js_1.default;
var checkTabbablePage_js_1 = __importDefault(require("./commands/checkTabbablePage.js"));
exports.checkTabbablePage = checkTabbablePage_js_1.default;
var constants_js_1 = require("./helpers/constants.js");
Object.defineProperty(exports, "IOS_OFFSETS", { enumerable: true, get: function () { return constants_js_1.IOS_OFFSETS; } });
Object.defineProperty(exports, "FOLDERS", { enumerable: true, get: function () { return constants_js_1.FOLDERS; } });
